### DISCUSSION TITLE
Tracking/saving fields outside of those involved in a PDE

### DISCUSSION THREAD

**Discussion by iamlll** [zwicker-group/py-pde#450] [Q&A] [answered]
Hello, 

I'm working with a system of (2D spatial) equations that looks something like this:

$$ \frac{\partial u}{\partial t} = \nabla^2 v$$

$$ \frac{\partial v}{\partial t} = -\nabla^2 u $$

I'd also like to keep track of quantities such as the derivatives of these fields (e.g. d_dx(u), d_dx(v)) even though they are not explicitly included in the PDEs to be solved, but could not find any examples in the py-PDE documentation to show how to do this. Is such a task currently possible, or should calculation of gradients be done in post-processing?

**Comment by david-zwicker** [comment]
You could use the `transformation` argument of the [`tracker` method of `FileStorage`](https://py-pde.readthedocs.io/en/latest/packages/pde.storage.base.html#pde.storage.base.StorageBase.tracker). Alternatively, you could of course re-create the derivatives from the stored fields in post-processing.

**Reply by iamlll** [reply]
Awesome, thank you so much! Just to clarify, does the callable function within `transformation` only take a single field argument (since the documentation says "The first argument always is the current field, while the optional second argument is the associated time"), or does it instead take the entire state of the system? So if for instance I wanted to keep track of both d_dx(u) and d_dx(v), I am guessing it should be possible to track them both at the same time, but am a little confused how to do so. Maybe something like

```
fun(state, t):
  dx = pde.grids.operators._make_derivative(grid,axis=0)
  du = dx(state['u'])
  dv = dx(state['v'])
  return du, dv
```
?

**Reply by david-zwicker** [reply]
The function gets the entire state (e.g., a `FieldCollection` if multiple fields are simulated). The function operators on fields and also needs to return a single field class (or a `FieldCollection` to return multiple fields). The current master branch has an addition where fields can be easily added to a field collection:

```python
fun(state, t):
    dx = pde.grids.operators._make_derivative(grid, axis=0)
    du = state['u'].apply(dx)
    dv = state['v'].apply(dx)
    return state.append(du, dv)
```

The `append` method only exists in master (currently). If you cannot update, you need to construct a `FieldCollection`. Note that the returned collection should also contain `state` itself since it otherwise would not be stored.

**Reply by iamlll** [reply]
I see - but when I run this code, I then get a `ValueError` since there are now 4 states the system must evolve forward in time but only 2 equations. Do you know of a workaround for this? I don't know if it makes any difference, but I constructed a new `FieldCollection` instead of appending to the original state.

**Reply by david-zwicker** [reply]
The idea is that you use the function `fun` in the `transformation` argument of the tracker,

```python
storage = FileStorage('test.hdf')
eq.solve(..., tracker=storager.tracker(1, transformation=fun))
```

**Reply by iamlll** [reply]
Yes - apologies, I was not clear in my previous response. I have been running
`result = eq.solve(state,t_range=tmax,dt=solvestep,tracker=writer.tracker(interval=dt,transformation=fun))`, 
which produces errors like
`ValueError: Expected 2 fields in state, but got 3 ones`

**Reply by david-zwicker** [reply]
Could you provide a self-contained example showing the error? This would help me to debug things!

**Reply by iamlll** [reply]
Of course, thank you so much!!
```
def demo():
    L = 20
    resolution = 5
    bc = "auto_periodic_neumann"
    grid = pde.CartesianGrid([[-int(L/2), int(L/2)], [-int(L/2),int(L/2)]], [int(resolution*L), int(resolution*L)], periodic=True)

    def fun(state, t):
        div = pde.ScalarField(grid,label='div')
        div.data = state['u']._apply_operator('d_dx',bc=bc) + state['v']._apply_operator('d_dy',bc=bc)
        newstate = state.fields
        newstate.append(div)
        return pde.FieldCollection(newstate)

    expr = {
        'u': f'laplace(v)',
        'v': f'-laplace(u)',
    }
    print(expr)
    u = pde.ScalarField(grid,data=0,label='u')
    v = pde.ScalarField(grid,data=0,label='v')
    eq = pde.PDE(expr,bc=bc)
    state = pde.FieldCollection([u,v])

    path = 'demo.hdf5'
    writer = pde.FileStorage(path,write_mode='truncate')
    result = eq.solve(state,t_range=10,dt=0.005,tracker=writer.tracker(interval=0.1,transformation=fun))
    return

if __name__ == '__main__':
    demo()
```

**Reply by david-zwicker** [reply, marked as answer]
Thank you for providing the code! I can reproduce the error, which is caused by a subtle behavior of py-pde. In essence, `new state.append(div)` actually also modifies `state.fields` and leads to the error. You could instead use

```python
newstate = state.fields[:]   # create a copy of the list
newstate.append(div)
```

or just

```python
newstate = state.fields[:] + [div]
```

to prevent this error!

**Reply by iamlll** [reply]
That worked for me, thank you so much!
